core/cc/android/crash_handler.cpp

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright (C) 2017 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "../crash_handler.h"
+
+#include "client/linux/handler/exception_handler.h"
+
+namespace {
+
+static
+bool handleCrash(const google_breakpad::MinidumpDescriptor& descriptor, void* crashHandlerPtr, bool succeeded) {
+    core::CrashHandler* crashHandler = reinterpret_cast<core::CrashHandler*>(crashHandlerPtr);
+    std::string minidumpPath(descriptor.path());
+    return crashHandler->handleMinidump(minidumpPath, succeeded);
+}
+
+}
+
+namespace core {
+
+CrashHandler::CrashHandler(HandlerFunction handlerFunction) :
+    mHandlerFunction(handlerFunction),
+    mHandler(new google_breakpad::ExceptionHandler(
+            google_breakpad::MinidumpDescriptor("./crashes"),
+            NULL, ::handleCrash, reinterpret_cast<void*>(this), true, -1)) {
+}
+
+// this prevents unique_ptr<CrashHandler> from causing an incomplete type error from inlining the destructor.
```

Comment:
eh? This is new to me. Can't you just put `~CrashHandler() = default;` in the header?

Reply:
No you can't, the unique_ptr needs to know the size of the object in it's destructor, which it can't know until the cpp file due to the forward declaration. Problem is if CrashHandler's default destructor is not manually set in a separate compilation unit it will be inlined by the compiler without knowledge of the size of google_breakpad::ExceptionHandler.
[This](https://stackoverflow.com/questions/13414652/forward-declaration-with-unique-ptr) and [this](https://stackoverflow.com/questions/6012157/is-stdunique-ptrt-required-to-know-the-full-definition-of-t) have more info on the issue.

Reply:
Fair enough - you didn't mention forward declaration in your comment.

Reply:
ah, I'll add that then.

Reply:
fixed.